File: Calto-GG/Microsoft/Day 1/Combination_Sum_III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> list[list[int]]:
        res=[]
        def recurse(i=0,c=1,r=[],s=0):
            if i==k:
                if s==n:
                    res.append(r[:])
            if i>=k or c>9:
                return
            r.append(c)
            recurse(i+1,c+1,r,s+c)
            r.pop()
            recurse(i,c+1,r,s)
            
        recurse()
        return res

File: Mayur/Microsoft/Day 1/Combination Sum III.cpp
// https://leetcode.com/problems/combination-sum-iii/solutions/3027004/easy-to-understand-backtracking-c/

class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>>res;
        vector<int>curr;
        getSubset(1,9,curr,res,k,n);
        return res;
    }
    void getSubset(int s,int e,vector<int>&curr,vector<vector<int>>&res,int k,int n)
    {
        if(n==0 and curr.size()==k)
        {
                 res.push_back(curr);
        }

        for(int i=s;i<=e;i++)
        {
            curr.push_back(i);
            getSubset(i+1,e,curr,res,k,n-i);
            curr.pop_back();
        }
    }
};
